libtsuba/src/RDGManifest.cpp

```diff
@@ -0,0 +1,320 @@
+using Result = katana::Result<T>;
+using json = nlohmann::json;
+
+namespace {
+
+Result<uint64_t>
+Parse(const std::string& str) {
+  uint64_t val = strtoul(str.c_str(), nullptr, 10);
+  if (errno == ERANGE) {
+    return KATANA_ERROR(
+        katana::ResultErrno(), "meta file found with out of range version");
+  }
+  return val;
+}
+
+katana::Result<std::vector<std::string>>
+FileList(const std::string& dir) {
+  std::vector<std::string> files;
+  auto list_fut = tsuba::FileListAsync(dir, &files);
+  KATANA_LOG_ASSERT(list_fut.valid());
+
+  if (auto res = list_fut.get(); !res) {
+    return res.error();
+  }
+  return files;
+}
+
+katana::Result<katana::Uri>
+FindLatestManifestFile(const katana::Uri& name) {
+  KATANA_LOG_DEBUG_ASSERT(!tsuba::RDGManifest::IsMetaUri(name));
+  auto list_res = FileList(name.string());
+  if (!list_res) {
+    return list_res.error();
+  }
+
+  uint64_t version = 0;
+  std::string found_meta;
+  for (const std::string& file : list_res.value()) {
+    if (auto res = tsuba::RDGManifest::ParseVersionFromName(file); res) {
+      uint64_t new_version = res.value();
+      if (new_version >= version) {
+        version = new_version;
+        found_meta = file;
+      }
+    }
+  }
+  if (found_meta.empty()) {
+    return KATANA_ERROR(
+        tsuba::ErrorCode::NotFound, "failed: could not find meta   file in {}",
+        name);
+  }
+  return name.Join(found_meta);
+}
+
+const int MANIFEST_MATCH_VERS_INDEX = 1;
+const int MANIFEST_MATCH_VIEW_INDEX = 2;
+
+}  // namespace
+
+namespace tsuba {
+
+const std::regex RDGManifest::kManifestVersion(
+    "katana_vers(?:([0-9]+))_(?:([0-9A-Za-z-]+))\\.manifest$");
+
+Result<tsuba::RDGManifest>
+RDGManifest::MakeFromStorage(const katana::Uri& uri) {
+  tsuba::FileView fv;
+
+  if (auto res = fv.Bind(uri.string(), true); !res) {
+    return res.error();
+    auto latest_meta_uri = FindLatestManifestFile(uri);
+    if (!latest_meta_uri) {
+      return res.error();
+    }
+    if (auto res_latest = fv.Bind(latest_meta_uri.value().string(), true);
+        !res_latest) {
+      return res.error();
+    }
+  }
+  tsuba::RDGManifest meta(uri.DirName());
+  auto meta_res = katana::JsonParse<tsuba::RDGManifest>(fv, &meta);
+
+  if (!meta_res) {
+    return meta_res.error().WithContext("cannot parse {}", uri.string());
+  }
+
+  auto manifest_name = uri.BaseName();
+  auto view_name = ParseViewNameFromName(manifest_name);
+  auto view_args = ParseViewArgsFromName(manifest_name);
+
+  if (view_name) {
+    meta.set_viewtype(view_name.value());
+  }
+
+  if (view_args) {
+    meta.set_viewargs(view_args.value());
+  } else {
+    meta.set_viewargs(std::vector<std::string>());
+  }
+  return meta;
+}
+
+Result<RDGManifest>
+RDGManifest::Make(
+    const katana::Uri& uri, const std::string& view_type, uint64_t version) {
+  return MakeFromStorage(FileName(uri, view_type, version));
+}
+
+Result<RDGManifest>
+RDGManifest::Make(RDGHandle handle) {
+  return handle.impl_->rdg_meta();
```

Comment:
another reference in a member name this time probably should be `->rdg_manifest()`

Reply:
Yeah, I have been meaning to go through and sed meta==>manifest. Will do that before completion.